File: python/snake.py
import sys, math, random, pygame
from pygame.locals import *

SPRITE_SIZE = 20
SCREEN_SIZE = 20
BLACK = 0, 0, 0
RED = 255, 0, 0
GREEN = 0, 255, 0
MIDDLE = math.floor(SCREEN_SIZE/2)
THIRD = math.floor(SCREEN_SIZE/3)
INITIAL_TAIL = 5
FRAMES_PER_SECOND = 15

def handle_input(speed):
    for event in pygame.event.get():
        if event.type == pygame.QUIT: 
            sys.exit()
        elif event.type == KEYDOWN:
            if speed.x == 0:
                if event.key == K_LEFT:
                    speed.x = -1
                    speed.y = 0
                elif event.key == K_RIGHT:
                    speed.x = 1
                    speed.y = 0
            if speed.y == 0:
                if event.key == K_UP:
                    speed.x = 0
                    speed.y = -1
                elif event.key == K_DOWN:
                    speed.x = 0
                    speed.y = 1
            if event.key == K_ESCAPE:
                sys.exit()              

def check_collision(obj, trail):
    for body in trail:
        if body == obj:
            return True
    return False

def generate_apple(apple, trail):
    while True:
        apple.x = random.randint(0, SCREEN_SIZE - 1)
        apple.y = random.randint(0, SCREEN_SIZE - 1)
        if check_collision(apple, trail) == False:
            break

def check_pickup(apple, player, trail, tail):
    if apple == player:
        tail += 1
        generate_apple(apple, trail)
    return tail

def update(player, apple, trail, tail, speed):
    player.x = (player.x + speed.x) % SCREEN_SIZE
    player.y = (player.y + speed.y) % SCREEN_SIZE

    if check_collision(player, trail):
        if tail > INITIAL_TAIL:
            print(tail - INITIAL_TAIL)
        tail = INITIAL_TAIL
        player.x = MIDDLE
        player.y = MIDDLE
        speed.x = 0
        speed.y = 0

    tail = check_pickup(apple, player, trail, tail)

    trail.append(pygame.Vector2(player.x, player.y))
    while len(trail) > tail:
        trail.pop(0)
    return tail

def draw(screen, player, apple, trail):
    # clear screen
    screen.fill(BLACK)

    # draw snake
    for body in trail:
        body_rect = Rect(body.x * SPRITE_SIZE + 1, body.y * SPRITE_SIZE + 1, SPRITE_SIZE - 1, SPRITE_SIZE - 1)
        pygame.draw.rect(screen, GREEN, body_rect)

    # draw apple
    apple_rect = Rect(apple.x * SPRITE_SIZE + 1, apple.y * SPRITE_SIZE + 1, SPRITE_SIZE - 1,  SPRITE_SIZE - 1)
    pygame.draw.rect(screen, RED, apple_rect)

def run():
    width = SCREEN_SIZE * SPRITE_SIZE
    height = SCREEN_SIZE * SPRITE_SIZE
    size = width, height 

    screen = pygame.display.set_mode(size)
    clock = pygame.time.Clock()

    player = pygame.Vector2(MIDDLE, MIDDLE)
    apple = pygame.Vector2(THIRD, THIRD)
    speed = pygame.Vector2()
    trail = []
    tail = INITIAL_TAIL

    pygame.init()
    while True:
        clock.tick(FRAMES_PER_SECOND)
        handle_input(speed)
        tail = update(player, apple, trail, tail, speed)
        draw(screen, player, apple, trail)
        pygame.display.flip()

run()

File: javascript/clean/snake.js
// variables
SPRITE_SIZE = 20
SCREEN_SIZE = 20
INITIAL_TAIL = 5
FPS = 15;

var surface;
var ctx;

let Apple = {
    x: 3,
    y: 3,

    width: SPRITE_SIZE - 1,
    height: SPRITE_SIZE - 1,
    color: 'red',

    draw: function() {
        ctx.fillStyle = this.color;
        ctx.fillRect(this.x * SPRITE_SIZE + 1, this.y * SPRITE_SIZE + 1, this.width, this.height);
    }
};

let Snake = {
    x: 10,
    y: 10,
    dx: 0,
    dy: 0,
    tail: INITIAL_TAIL,
    trail: [],

    width: SPRITE_SIZE - 1,
    height: SPRITE_SIZE - 1,
    color: 'green',

    checkCollision: function() {
        this.trail.forEach(element => {
            if(element.x == this.x && element.y == this.y) {
                if(snake.tail > INITIAL_TAIL){
                    console.log("Scored: %i", this.tail - INITIAL_TAIL);
                }
                this.x = this.y = 10;
                this.dx = this.dy = 0;
                this.tail = INITIAL_TAIL;
            }
        });
    },

    update: function() {
        this.x = (this.x + this.dx + SCREEN_SIZE) % SCREEN_SIZE;
        this.y = (this.y + this.dy + SCREEN_SIZE) % SCREEN_SIZE;

        this.checkCollision();

        this.trail.push({x:this.x, y:this.y});
        while(this.trail.length > this.tail){
            this.trail.shift();
        }
    },

    draw: function() {
        ctx.fillStyle = this.color;
        this.trail.forEach(element => {
            ctx.fillRect(element.x * SPRITE_SIZE + 1, element.y * SPRITE_SIZE + 1, this.width, this.height);
        });
    }
};

let game = {
    snake: Snake,
    apple: Apple,

    generateApple: function() {
        needNewApple = false;
        do {
            needNewApple = false;
            this.apple.x = Math.floor(Math.random() * SCREEN_SIZE);
            this.apple.y = Math.floor(Math.random() * SCREEN_SIZE);
            this.snake.trail.forEach(element => {
                needNewApple |= (element.x == this.apple.x && element.y == this.apple.y);
            });
        } while(needNewApple)
    },

    checkPickup: function() {
        if(this.apple.x == this.snake.x && this.apple.y == this.snake.y){
            this.snake.tail++;
            this.generateApple();
        }
    },

    update: function() {
        this.snake.update();
        this.checkPickup();
    },

    draw: function() {
        ctx.fillStyle = "black";
        ctx.fillRect(0, 0, surface.width, surface.height);

        this.snake.draw();
        this.apple.draw();
    },
}

function keyPush(evt){
    if(game.snake.dx == 0) {
        switch(evt.keyCode){
            case 37:
                game.snake.dx = -1;
                game.snake.dy = 0;
                break;
            case 39:
                game.snake.dx = 1;
                game.snake.dy = 0;
                break;
        }
    }
    if(game.snake.dy == 0){
        switch(evt.keyCode){
            case 38:
                game.snake.dx = 0;
                game.snake.dy = -1;
                break;
            case 40:
                game.snake.dx = 0;
                game.snake.dy = 1;
                break;
        }
    }
}

function gameLoop() {
    game.update();
    game.draw();
}

window.onload = function() {
    surface = document.getElementById("surface");
    ctx = surface.getContext("2d");
    document.addEventListener("keydown", keyPush);
    setInterval(gameLoop, 1000 / FPS);
}
